### PR TITLE
[EPROD-437] Update reserve-address endpoint

### PR DESCRIPTION
**Pull request by itsyaasir**
# Issue ticket

Issue ticket link: <>

### Code conventions

- [x] I have performed a self-review of my code
- [x] Every new function is documented
- [x] Object names are auto explicative

### Security

- [x] The PR does not break APIs backward compatibility
- [x] The PR does not break the stable storage backward compatibility

### Testing

- [x] Every function is properly unit tested
- [x] I have added integration tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] IC endpoints are always tested through the `canister_call!` macro

### CODE REVIEW

File: src/register-evm-agent/src/register_evm_agent/src/reservation.rs [hunk 1]
```diff
@@ -62,8 +64,26 @@ impl<'a> ReservationService<'a> {
...
+            chain_id: DEFAULT_CHAIN_ID,
+        }
+        .calculate_hash_and_build()?;
+
+        info!("sending transaction to reserve address...");
+        let tx_hash = self.client.send_raw_transaction(tx).await??;
+
         self.client
-            .reserve_address(self.reserve_canister_id, address)
+            .reserve_address(self.reserve_canister_id, tx_hash)
```
Comment (ufoscout) [inline src/register-evm-agent/src/register_evm_agent/src/reservation.rs:86]: before reserving you should wait for the TX to be processed. For example, by getting the status in a loop every second

**Review by ufoscout — CHANGES_REQUESTED** [review]

File: src/register-evm-agent/src/register_evm_agent/src/reservation.rs (hunk 1 above)
Comment (itsyaasir) [reply src/register-evm-agent/src/register_evm_agent/src/reservation.rs:86]: ah, Yes

**Review by ufoscout — APPROVED** [review]

File: src/register-evm-agent/src/register_evm_agent/src/reservation.rs [hunk 2]
```diff
@@ -103,4 +135,25 @@ impl<'a> ReservationService<'a> {
...
+            info!("waiting for transaction to be finalized...");
+            tokio::time::sleep(std::time::Duration::from_secs(1)).await;
+
+            let tx_receipt = self
+                .client
+                .eth_get_transaction_receipt(tx_hash.clone())
+                .await??;
+
+            if tx_receipt.is_some() {
+                return Ok(());
```
Comment (GraDKh) [inline src/register-evm-agent/src/register_evm_agent/src/reservation.rs:153]: Transaction may be executed but have status 0. Are we ok with that?

File: src/register-evm-agent/src/register_evm_agent/src/reservation.rs [hunk 3]
```diff
@@ -103,4 +135,25 @@ impl<'a> ReservationService<'a> {
...
         Ok(())
     }
+
+    /// Wait for the transaction to be finalized
+    async fn wait_for_transaction(&self, tx_hash: H256) -> Result<()> {
+        const MAX_RETRIES: u32 = 10;
+
+        for _ in 0..MAX_RETRIES {
+            info!("waiting for transaction to be finalized...");
+            tokio::time::sleep(std::time::Duration::from_secs(1)).await;
```
Comment (GraDKh) [inline src/register-evm-agent/src/register_evm_agent/src/reservation.rs:145]: Does tokio::sleep work in ic runtime?

Comment (itsyaasir) [reply src/register-evm-agent/src/register_evm_agent/src/reservation.rs:145]: This is not ic runtime :)

File: src/register-evm-agent/src/register_evm_agent/src/reservation.rs (hunk 2 above)
Comment (itsyaasir) [reply src/register-evm-agent/src/register_evm_agent/src/reservation.rs:153]: Prolly not!, good catch

File: src/register-evm-agent/src/register_evm_agent/src/reservation.rs [hunk 4]
```diff
@@ -103,4 +135,27 @@ impl<'a> ReservationService<'a> {
...
+
+            let tx_receipt = self
+                .client
+                .eth_get_transaction_receipt(tx_hash.clone())
+                .await??;
+
+            if let Some(receipt) = tx_receipt {
+                if receipt.status == Some(1_u64.into()) {
```
Comment (GraDKh) [inline src/register-evm-agent/src/register_evm_agent/src/reservation.rs:153]: I think that in opposite case we should return error immediately, there is no point to continue the cycle

**Review by GraDKh — APPROVED** [review]

**Pull request merged**